thali/thalireplicationmanager.js

```diff
@@ -33,30 +43,93 @@ ThaliReplicationManager.events = {
   START_ERROR: 'startError',
   STOP_ERROR: 'stopError',
   CONNECT_ERROR: 'connectError',
-  DISCONNECT_ERROR: 'disconnectError',
-  SYNC_ERROR: 'syncRetry'
+  DISCONNECT_ERROR: 'disconnectError'
 };
 
 /**
-* Starts the Thali replication manager with the given device name and port number
-* @param {String} deviceName the device name to advertise.
+* Returns the current device's id if available. If not, the provided callback
+* function is saved and called when the id does become available.
+* @param {Function} cb the callback which returns the current device's id.
+*/
+ThaliReplicationManager.prototype.getDeviceIdentity = function (cb) {
+  if(this._deviceIdentityFlag == deviceIdentityFlag.deviceIdentityAvailable) {
+    cb(null, this._deviceName);
+    return;
+  }
+  if(this._deviceIdentityFlag == deviceIdentityFlag.noDeviceIdentitySet) {
+    this._deviceIdentityFlag = deviceIdentityFlag.gettingDeviceIdentity;
+    getPublicKeyHash.call(this, cb);
+    return;
+  }
+  if(this._deviceIdentityFlag == deviceIdentityFlag.gettingDeviceIdentity) {
+    // save the callback for future
+    this._deviceIdentityListeners.push(cb);
+  }
+}
+
+function getPublicKeyHash(cb) {
+  cryptomanager.getPublicKeyHash(function (err, publicKeyHash) {
+    if (err) {
+      this._deviceIdentityFlag = deviceIdentityFlag.noDeviceIdentitySet;
+      informDeviceIdentityListeners.call(this, err, null);
+      this._deviceIdentityListeners = [];
+      cb(err);
+    } else {
+      this._deviceName = publicKeyHash;
+      this._deviceIdentityFlag = deviceIdentityFlag.deviceIdentityAvailable;
+      informDeviceIdentityListeners.call(this, null, publicKeyHash);
+      this._deviceIdentityListeners = [];
+      cb(null, publicKeyHash);
+    }
+  }.bind(this));
+}
+
+function informDeviceIdentityListeners(err, publicKeyHash) {
+  this._deviceIdentityListeners.forEach(function (listener) {
+    listener(err, publicKeyHash);
+  });
+}
+
+/**
+* Starts the Thali replication manager with the given port number and db name.
+* The device-id is obtained using the cryptomanager's API.
 * @param {Number} port the port number used for synchronization.
+* @param {String} dbName the name of the database.
 */
-ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
-  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+ThaliReplicationManager.prototype.start = function (port, dbName) {
   validations.ensureValidPort(port);
   validations.ensureNonNullOrEmptyString(dbName, 'dbName');
+  
+  // get the device identity first if the user did not do it
+  if(!this._deviceName) {
```

Comment:
Now your state machine is stretched across two different places, getDeviceIdentity and here. Why would you do that? It just increases the probability of screwing something up. Wouldn't it be simpler to just call getDeviceIdentity and in the callback call startReplicationManager? Since start doesn't have a return value (it instead emits an event when things get going) it doesn't matter that you point an asynchronous call in here and it simplifies the heck out of your code.